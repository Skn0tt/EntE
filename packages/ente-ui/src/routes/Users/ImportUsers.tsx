/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import { AppState, createUsersRequest, getStudents } from "../../redux";
import { Button, Dialog, Grid } from "@material-ui/core";
import withMobileDialog from "@material-ui/core/withMobileDialog";
import DialogActions from "@material-ui/core/DialogActions";
import * as React from "react";
import Dropzone from "react-dropzone";
import { connect, MapStateToPropsParam } from "react-redux";
import { Action, Dispatch } from "redux";
import { CreateUserDto } from "ente-types";
import { parseCSVFromFile } from "../../helpers/parser";
import SignedAvatar from "../../elements/SignedAvatar";
import { createTranslation } from "../../helpers/createTranslation";
import { MessagesContext, MessagesContextValue } from "../../context/Messages";

const lang = createTranslation({
  en: {
    submit: "Import",
    close: "Close",
    dropzone: "Drop a .csv file here."
  },
  de: {
    submit: "Importieren",
    close: "SchlieÃŸen",
    dropzone: "Legen sie hier eine .csv-Datei ab."
  }
});

const readFile = async (f: File) =>
  new Promise<string>((resolve, reject) => {
    const r = new FileReader();
    r.onload = _ => resolve(r.result as string);
    r.onerror = reject;
    r.readAsText(f);
  });

/**
 * # Component Types
 */
interface OwnProps {
  onClose(): void;
  show: boolean;
}

interface StateProps {
  usernames: string[];
}
const mapStateToProps: MapStateToPropsParam<
  StateProps,
  OwnProps,
  AppState
> = state => ({
  usernames: getStudents(state).map(u => u.get("username"))
});

interface DispatchProps {
  createUsers(users: CreateUserDto[]): void;
}
const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  createUsers: (users: CreateUserDto[]) => dispatch(createUsersRequest(users))
});

interface InjectedProps {
  fullScreen: boolean;
}

interface State {
  users: CreateUserDto[];
  error: boolean;
}

type Props = OwnProps & DispatchProps & StateProps;

/**
 * # Component
 */
export class ImportUsers extends React.Component<Props & InjectedProps, State> {
  /**
   * ## Intialization
   */
  state: State = {
    users: [],
    error: true
  };

  static contextType = MessagesContext;

  /**
   * # Handlers
   */
  handleClose = () => this.props.onClose();

  handleSubmit = () =>
    this.state.users.length !== 0 && this.props.createUsers(this.state.users);

  onDrop = async (accepted: File[], rejected: File[]) => {
    if (accepted.length === 0) {
      return;
    }

    try {
      const input = await readFile(accepted[0]);
      const users = await parseCSVFromFile(input, this.props.usernames);
      this.setState({ users, error: false });
    } catch (error) {
      this.setState({ error: true });
      const { addMessages } = this.context as MessagesContextValue;
      addMessages(error.message);
    }
  };

  /**
   * # Validation
   */
  inputValid = (): boolean =>
    !this.state.error && this.state.users.length !== 0;

  /**
   * # Render
   */
  render() {
    const { show, fullScreen } = this.props;
    const { error } = this.state;

    return (
      <Dialog fullScreen={fullScreen} onClose={this.handleClose} open={show}>
        <Grid container direction="column">
          <Grid item xs={12}>
            <Dropzone
              accept="text/csv"
              onDrop={this.onDrop}
              className="dropzone"
            >
              {lang.dropzone}
            </Dropzone>
          </Grid>
          <Grid item xs={12}>
            <SignedAvatar signed={!error} />
          </Grid>
        </Grid>
        <DialogActions>
          <Button
            onClick={this.handleClose}
            color="secondary"
            className="close"
          >
            {lang.close}
          </Button>
          <Button
            onClick={() => {
              this.handleSubmit();
              this.handleClose();
            }}
            disabled={!this.inputValid()}
            color="primary"
            className="submit"
          >
            {lang.submit}
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

export default connect<StateProps, DispatchProps, OwnProps, AppState>(
  mapStateToProps,
  mapDispatchToProps
)(withMobileDialog<Props>()(ImportUsers));
