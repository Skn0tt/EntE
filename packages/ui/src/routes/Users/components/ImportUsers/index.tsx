import { parseCSVFromFile } from "ente-parser";
import {
  AppState,
  addMessage,
  createUsersRequest,
  getUsers,
  getStudents
} from "ente-redux";
import { IUserCreate } from "ente-types";
import { Button, Dialog, Grid } from "material-ui";
import { withMobileDialog } from "material-ui/Dialog";
import DialogActions from "material-ui/Dialog/DialogActions";
import withStyles, { WithStyles } from "material-ui/styles/withStyles";
import * as React from "react";
import Dropzone from "react-dropzone";
import { connect } from "react-redux";
import { Action, Dispatch } from "redux";
import SignedAvatar from "../../../SpecificEntry/elements/SignedAvatar";
import UnsignedAvatar from "../../../SpecificEntry/elements/UnsignedAvatar";
import styles from "./styles";
import lang from "ente-lang";

/**
 * # Component Types
 */
interface OwnProps {
  onClose(): void;
  show: boolean;
}

interface StateProps {
  usernames: string[];
}
const mapStateToProps = (state: AppState): StateProps => ({
  usernames: getStudents(state).map(u => u.get("username"))
});

interface DispatchProps {
  createUsers(users: IUserCreate[]): void;
  addMessage(msg: string): void;
}
const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  createUsers: (users: IUserCreate[]) => dispatch(createUsersRequest(users)),
  addMessage: (msg: string) => dispatch(addMessage(msg))
});

interface InjectedProps {
  fullScreen: boolean;
}

interface State {
  users: IUserCreate[];
  error: boolean;
}

type Props = OwnProps & DispatchProps & StateProps & WithStyles & InjectedProps;

/**
 * # Component
 */
export class ImportUsers extends React.Component<Props, State> {
  /**
   * ## Intialization
   */
  state: State = {
    users: [],
    error: true
  };

  /**
   * # Handlers
   */
  handleClose = () => this.props.onClose();

  handleSubmit = () =>
    this.state.users.length !== 0 && this.props.createUsers(this.state.users);

  onDrop = async (accepted: File[], rejected: File[]) => {
    if (accepted.length === 0) {
      return;
    }

    try {
      const users = await parseCSVFromFile(accepted[0], this.props.usernames);
      this.setState({ users, error: false });
    } catch (error) {
      this.setState({ error: true });
      this.props.addMessage(error.message);
    }
  };

  /**
   * # Validation
   */
  inputValid = (): boolean =>
    !this.state.error && this.state.users.length !== 0;

  /**
   * # Render
   */
  render() {
    const { show, fullScreen } = this.props;
    const { error } = this.state;

    return (
      <Dialog fullScreen={fullScreen} onClose={this.handleClose} open={show}>
        <Grid container direction="column">
          <Grid item xs={12}>
            <Dropzone
              accept="text/csv"
              onDrop={this.onDrop}
              className="dropzone"
            >
              {lang().ui.importUsers.dropzone}
            </Dropzone>
          </Grid>
          <Grid item xs={12}>
            {error ? <UnsignedAvatar /> : <SignedAvatar />}
          </Grid>
        </Grid>
        <DialogActions>
          <Button
            onClick={this.handleClose}
            color="secondary"
            className="close"
          >
            {lang().ui.common.close}
          </Button>
          <Button
            onClick={() => {
              this.handleSubmit();
              this.handleClose();
            }}
            disabled={!this.inputValid()}
            color="primary"
            className="submit"
          >
            {lang().ui.common.submit}
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

export default connect<StateProps, DispatchProps>(
  mapStateToProps,
  mapDispatchToProps
)(
  withMobileDialog<OwnProps & DispatchProps & StateProps>()(
    withStyles(styles)(ImportUsers)
  )
);
