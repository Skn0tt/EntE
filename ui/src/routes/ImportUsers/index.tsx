import * as React from 'react';
import styles from './styles';
import withStyles, { WithStyles } from 'material-ui/styles/withStyles';
import { Dialog, Grid, Button } from 'material-ui';
import { withMobileDialog } from 'material-ui/Dialog';
import { connect } from 'react-redux';
import { AppState, IUserCreate } from '../../interfaces/index';
import { Dispatch, Action } from 'redux';
import { RouteComponentProps } from 'react-router';
// tslint:disable-next-line:import-name
import Dropzone from 'react-dropzone';
import DialogActions from 'material-ui/Dialog/DialogActions';
import parseCSV from './services/parseCSV';
import { createUserRequest, addMessage } from '../../redux/actions';
import UnsignedAvatar from '../SpecificEntry/elements/UnsignedAvatar';
import SignedAvatar from '../SpecificEntry/elements/SignedAvatar';

interface StateProps {}
const mapStateToProps = (state: AppState) => ({});

interface DispatchProps {
  createUsers(users: IUserCreate[]): void;
  addMessage(msg: string): void;
}
const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  createUsers: (users: IUserCreate[]) => dispatch(createUserRequest(users)),
  addMessage: (msg: string) => dispatch(addMessage(msg)),
});

interface InjectedProps {
  fullScreen: boolean;
}

interface State {
  users: IUserCreate[];
  error: boolean;
}

type Props = DispatchProps & StateProps & WithStyles & InjectedProps & RouteComponentProps<{}>;

const ImportUsers =
  connect<StateProps, DispatchProps>(mapStateToProps, mapDispatchToProps)(
  withStyles(styles)(
  withMobileDialog<Props>()(
class extends React.Component<Props, State> {
  state: State = {
    users: [],
    error: true,
  };

  onDrop = async (accepted: File[], rejected: File[]) => {
    try {
      const users = await parseCSV(accepted[0]);
      this.setState({Â users, error: false });
    } catch (error) {
      this.setState({ error: true });
      this.props.addMessage(error.message);
    }
  }

  handleClose = () => this.props.history.goBack();
  handleSubmit = async () => {
    this.props.createUsers(this.state.users);
  }

  /**
   * # Validation
   */
  inputValid = (): boolean => !this.state.error;
  
  render() {
    const { props, state } = this;
    return (
      <Dialog
        fullScreen={props.fullScreen}
        onClose={() => props.history.goBack()}
        open
      >
        <Grid container direction="column">
          <Grid item xs={12}>
            <Dropzone
              accept="text/csv"
              onDrop={this.onDrop}
            >
              Drop items here!
            </Dropzone>
          </Grid>
          <Grid item xs={12}>
            {state.error ? <UnsignedAvatar /> : <SignedAvatar />}
          </Grid>
        </Grid>
        <DialogActions>
          <Button onClick={this.handleClose} color="secondary">
            Cancel
          </Button>
          <Button
            onClick={() => {
              this.handleSubmit();
              this.handleClose();
            }}
            disabled={!this.inputValid()}
            color="primary"
          >
            OK
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
})));

export default ImportUsers;
