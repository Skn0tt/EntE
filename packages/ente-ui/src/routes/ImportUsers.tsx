/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import {
  Button,
  createStyles,
  Dialog,
  Grid,
  Theme,
  Typography,
  withStyles,
  WithStyles,
  DialogTitle,
  DialogContent,
  DialogContentText
} from "@material-ui/core";
import DialogActions from "@material-ui/core/DialogActions";
import withMobileDialog, {
  InjectedProps
} from "@material-ui/core/withMobileDialog";
import { ValidationError } from "class-validator";
import { CreateUserDto } from "ente-types";
import { Maybe, None, Some } from "monet";
import * as React from "react";
import Dropzone from "react-dropzone";
import {
  connect,
  MapStateToPropsParam,
  MapDispatchToPropsParam
} from "react-redux";
import { parseCSVFromFile } from "../helpers/parser";
import { AppState, getStudents, UserN, importUsersRequest } from "../redux";
import * as _ from "lodash";
import ErrorDisplay from "./Users/ErrorDisplay";
import { UserTable } from "./Users/UserTable";
import { makeTranslationHook } from "../helpers/makeTranslationHook";
import { CheckboxWithDescription } from "../components/CheckboxWithDescription";

const useTranslation = makeTranslationHook({
  en: {
    submit: "Import",
    title: "Import",
    close: "Close",
    dropzone: "Drop a .csv file or click here.",
    pleaseUploadCsvFile: "Please upload a .csv file.",
    deleteEntries: {
      title: "Delete all entries",
      caption:
        "Delete all entries and the corresponding slots, for example at the start of a term."
    },
    deleteUsers: {
      title: "Delete all users",
      caption:
        "Delete all users that are not contained nor referenced by the import, for example at the start of a term."
    },
    description: `
      Users whose usernames are already present in the database are updated using the data from the import.
      New users are created.
      If a new user's import data contains no password, he will receive an invitation email.

      Beware: In order for the update to work, usernames need to be consistent with the last import.
    `.trim()
  },
  de: {
    submit: "Importieren",
    title: "Importieren",
    close: "Schließen",
    dropzone: "Legen Sie eine .csv-Datei ab oder klicken Sie hier.",
    pleaseUploadCsvFile: "Bitte laden Sie eine .csv-Datei hoch.",
    deleteEntries: {
      title: "Alle Einträge löschen",
      caption:
        "Alle bestehenden Einträge sowie die zugehörigen Stunden löschen, zum Beispiel am Schuljahresbeginn."
    },
    deleteUsers: {
      title: "Alle Nutzer löschen",
      caption:
        "Alle Nutzer löschen, die nicht im Import enthalten sind oder referenziert werden, zum Beispiel am Schuljahresbeginn."
    },
    description: `
      Nutzer, deren Nutzernamen schon in der Datenbank enthalten sind, werden mit den Daten aus dem Import aktualisiert.
      Neue Nutzer werden erstellt.
      Enthalten die Import-Daten eines neuen Nutzers kein Passwort, so erhält er eine Einladungs-Email.

      Achtung: Damit die Aktualisierung fehlerlos funktionieren kann, muss die Vergabe der Nutzernamen identisch zum letzten Import sein.
    `.trim()
  }
});

const toUserN = (u: CreateUserDto): UserN => {
  return new UserN({
    username: u.username,
    childrenIds: [],
    email: u.email,
    displayname: u.displayname,
    graduationYear: u.graduationYear,
    birthday: u.birthday,
    id: u.username,
    role: u.role
  });
};

const readFile = async (f: File) => {
  const response = new Response(f);
  return await response.text();
};

/**
 * # Component Types
 */
interface ImportUsersOwnProps {
  onClose(): void;
  show: boolean;
}

interface ImportUsersStateProps {
  usernames: string[];
}
const mapStateToProps: MapStateToPropsParam<
  ImportUsersStateProps,
  ImportUsersOwnProps,
  AppState
> = state => ({
  usernames: getStudents(state).map(u => u.get("username"))
});

interface ImportUsersDispatchProps {
  importUsers: (
    dtos: CreateUserDto[],
    deleteEntries: boolean,
    deleteUsers: boolean
  ) => void;
}
const mapDispatchToProps: MapDispatchToPropsParam<
  ImportUsersDispatchProps,
  ImportUsersOwnProps
> = dispatch => ({
  importUsers: (dtos, deleteEntries, deleteUsers) =>
    dispatch(importUsersRequest({ dtos, deleteEntries, deleteUsers }))
});

type ImportUsersProps = ImportUsersOwnProps &
  ImportUsersStateProps &
  ImportUsersDispatchProps &
  WithStyles<"dropzone"> &
  InjectedProps;

const ImportUsers: React.FunctionComponent<ImportUsersProps> = props => {
  const { fullScreen, show, classes, onClose, usernames, importUsers } = props;
  const translation = useTranslation();

  const [deleteEntries, setDeleteEntries] = React.useState(false);
  const [deleteUsers, setDeleteUsers] = React.useState(false);
  const [users, setUsers] = React.useState<Maybe<CreateUserDto[]>>(None());
  const [errors, setErrors] = React.useState<
    Maybe<(ValidationError | string)[]>
  >(None());

  const handleSubmit = React.useCallback(
    () => {
      users.forEach(dtos => importUsers(dtos, deleteEntries, deleteUsers));
    },
    [users, importUsers, deleteEntries, deleteUsers]
  );

  const onDrop = async (accepted: File[]) => {
    const [file] = accepted;

    if (!file) {
      return;
    }

    const input = await readFile(file);
    const result = await parseCSVFromFile(input, usernames);
    result.forEach(success => {
      setUsers(Some(success));
      setErrors(None());
    });
    result.forEachFail(fail => {
      setUsers(None());
      setErrors(Some(fail));
    });
  };

  const inputIsValid = users.isSome();

  return (
    <Dialog fullScreen={fullScreen} onClose={onClose} open={show}>
      <DialogTitle>{translation.title}</DialogTitle>
      <DialogContent>
        <Grid container spacing={24} direction="column">
          <Grid item xs={12}>
            <DialogContentText>{translation.description}</DialogContentText>
          </Grid>
          <Grid item xs={12}>
            <Grid container spacing={16} direction="column">
              <Grid item xs={12}>
                <CheckboxWithDescription
                  title={translation.deleteUsers.title}
                  caption={translation.deleteUsers.caption}
                  onChange={setDeleteUsers}
                />
              </Grid>
              <Grid item xs={12}>
                <CheckboxWithDescription
                  title={translation.deleteEntries.title}
                  caption={translation.deleteEntries.caption}
                  onChange={setDeleteEntries}
                />
              </Grid>
            </Grid>
          </Grid>
          <Grid item xs={12}>
            <Dropzone
              onDrop={onDrop}
              className={classes.dropzone}
              accept=".csv"
            >
              <Typography variant="body1">{translation.dropzone}</Typography>
            </Dropzone>
          </Grid>
          {users
            .map(users => (
              <Grid item xs={12}>
                <UserTable users={users.map(toUserN)} />
              </Grid>
            ))
            .orSome(<></>)}
          {errors
            .map(errors => (
              <Grid item xs={12}>
                <ErrorDisplay errors={errors} />
              </Grid>
            ))
            .orSome(<></>)}
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose} color="secondary" className="close">
          {translation.close}
        </Button>
        <Button
          onClick={handleSubmit}
          disabled={!inputIsValid}
          color="primary"
          className="submit"
        >
          {translation.submit}
        </Button>
      </DialogActions>
    </Dialog>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    dropzone: {
      minHeight: 24,
      border: `1px solid ${theme.palette.grey[300]}`,
      borderRadius: theme.spacing.unit,
      padding: theme.spacing.unit * 2,
      boxSizing: "border-box"
    }
  });

export default connect<
  ImportUsersStateProps,
  ImportUsersDispatchProps,
  ImportUsersOwnProps,
  AppState
>(
  mapStateToProps,
  mapDispatchToProps
)(withStyles(styles)(withMobileDialog<ImportUsersProps>()(ImportUsers)));
