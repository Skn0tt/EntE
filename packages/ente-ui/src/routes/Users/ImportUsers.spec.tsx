/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import * as React from "react";
import { ImportUsers } from "./ImportUsers";
import { shallow } from "enzyme";

describe.skip("ImportUsers", () => {
  const onClose = jest.fn();
  const onImport = jest.fn();
  const comp = shallow(
    <ImportUsers
      show
      onImport={onImport}
      classes={{ dropzone: "dropzone" }}
      onClose={onClose}
      fullScreen
      usernames={[]}
    />
  );

  it("renders correctly", () => {
    const comp = shallow(
      <ImportUsers
        show
        classes={{ dropzone: "dropzone" }}
        onClose={onClose}
        onImport={onImport}
        fullScreen
        usernames={[]}
      />
    );

    expect(comp).toMatchSnapshot();
  });

  it("closes on clicking close", () => {
    comp.find(".close").simulate("click");
    expect(onClose).toHaveBeenCalled();
  });

  describe("when not added users", () => {
    it("doesn't submit", () => {
      comp.find(".submit").simulate("click");
      expect(onImport).not.toHaveBeenCalled();
    });
  });
});
