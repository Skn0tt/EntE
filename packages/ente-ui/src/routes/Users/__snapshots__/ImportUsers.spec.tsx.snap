// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportUsers renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImportUsers
    addMessage={[MockFunction]}
    createUsers={[MockFunction]}
    fullScreen={true}
    onClose={[MockFunction]}
    show={true}
    usernames={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <WithStyles(Grid)
          container={true}
          direction="column"
        >
          <WithStyles(Grid)
            item={true}
            xs={12}
          >
            <UNDEFINED
              accept="text/csv"
              className="dropzone"
              onDrop={[Function]}
            >
              Drop a .csv file here.
            </UNDEFINED>
          </WithStyles(Grid)>
          <WithStyles(Grid)
            item={true}
            xs={12}
          >
            <WithStyles(SignedAvatar)
              signed={false}
            />
          </WithStyles(Grid)>
        </WithStyles(Grid)>,
        <WithStyles(DialogActions)>
          <WithStyles(Button)
            className="close"
            color="secondary"
            onClick={[Function]}
          >
            Close
          </WithStyles(Button)>
          <WithStyles(Button)
            className="submit"
            color="primary"
            disabled={true}
            onClick={[Function]}
          >
            Import
          </WithStyles(Button)>
        </WithStyles(DialogActions)>,
      ],
      "fullScreen": true,
      "onClose": [Function],
      "open": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(Grid)
              item={true}
              xs={12}
            >
              <UNDEFINED
                accept="text/csv"
                className="dropzone"
                onDrop={[Function]}
              >
                Drop a .csv file here.
              </UNDEFINED>
            </WithStyles(Grid)>,
            <WithStyles(Grid)
              item={true}
              xs={12}
            >
              <WithStyles(SignedAvatar)
                signed={false}
              />
            </WithStyles(Grid)>,
          ],
          "container": true,
          "direction": "column",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <UNDEFINED
                accept="text/csv"
                className="dropzone"
                onDrop={[Function]}
              >
                Drop a .csv file here.
              </UNDEFINED>,
              "item": true,
              "xs": 12,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "accept": "text/csv",
                "children": "Drop a .csv file here.",
                "className": "dropzone",
                "onDrop": [Function],
              },
              "ref": null,
              "rendered": "Drop a .csv file here.",
              "type": undefined,
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <WithStyles(SignedAvatar)
                signed={false}
              />,
              "item": true,
              "xs": 12,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "signed": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(Button)
              className="close"
              color="secondary"
              onClick={[Function]}
            >
              Close
            </WithStyles(Button)>,
            <WithStyles(Button)
              className="submit"
              color="primary"
              disabled={true}
              onClick={[Function]}
            >
              Import
            </WithStyles(Button)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Close",
              "className": "close",
              "color": "secondary",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Close",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Import",
              "className": "submit",
              "color": "primary",
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Import",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Grid)
            container={true}
            direction="column"
          >
            <WithStyles(Grid)
              item={true}
              xs={12}
            >
              <UNDEFINED
                accept="text/csv"
                className="dropzone"
                onDrop={[Function]}
              >
                Drop a .csv file here.
              </UNDEFINED>
            </WithStyles(Grid)>
            <WithStyles(Grid)
              item={true}
              xs={12}
            >
              <WithStyles(SignedAvatar)
                signed={false}
              />
            </WithStyles(Grid)>
          </WithStyles(Grid)>,
          <WithStyles(DialogActions)>
            <WithStyles(Button)
              className="close"
              color="secondary"
              onClick={[Function]}
            >
              Close
            </WithStyles(Button)>
            <WithStyles(Button)
              className="submit"
              color="primary"
              disabled={true}
              onClick={[Function]}
            >
              Import
            </WithStyles(Button)>
          </WithStyles(DialogActions)>,
        ],
        "fullScreen": true,
        "onClose": [Function],
        "open": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <WithStyles(Grid)
                item={true}
                xs={12}
              >
                <UNDEFINED
                  accept="text/csv"
                  className="dropzone"
                  onDrop={[Function]}
                >
                  Drop a .csv file here.
                </UNDEFINED>
              </WithStyles(Grid)>,
              <WithStyles(Grid)
                item={true}
                xs={12}
              >
                <WithStyles(SignedAvatar)
                  signed={false}
                />
              </WithStyles(Grid)>,
            ],
            "container": true,
            "direction": "column",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <UNDEFINED
                  accept="text/csv"
                  className="dropzone"
                  onDrop={[Function]}
                >
                  Drop a .csv file here.
                </UNDEFINED>,
                "item": true,
                "xs": 12,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "accept": "text/csv",
                  "children": "Drop a .csv file here.",
                  "className": "dropzone",
                  "onDrop": [Function],
                },
                "ref": null,
                "rendered": "Drop a .csv file here.",
                "type": undefined,
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <WithStyles(SignedAvatar)
                  signed={false}
                />,
                "item": true,
                "xs": 12,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "signed": false,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <WithStyles(Button)
                className="close"
                color="secondary"
                onClick={[Function]}
              >
                Close
              </WithStyles(Button)>,
              <WithStyles(Button)
                className="submit"
                color="primary"
                disabled={true}
                onClick={[Function]}
              >
                Import
              </WithStyles(Button)>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Close",
                "className": "close",
                "color": "secondary",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Close",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Import",
                "className": "submit",
                "color": "primary",
                "disabled": true,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Import",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
